File: 54. Spiral Matrix/spiralOrder.js
/**
 * @param {number[][]} matrix
 * @return {number[]}
 */
const spiralOrder = (matrix) => {
  let left = 0;
  let right = matrix[0].length;
  let top = 0;
  let bottom = matrix.length;

  const res = [];

  while (left < right && top < bottom) {
    // left -> right
    for (let i = left; i < right; i++) {
      res.push(matrix[top][i]);
    }
    top++;

    // top -> bottom
    for (let i = top; i < bottom; i++) {
      res.push(matrix[i][right - 1]);
    }
    right--;

    if (!(left < right && top < bottom)) {
      break;
    }

    // right -> left
    for (let i = right - 1; i > left - 1; i--) {
      res.push(matrix[bottom - 1][i]);
    }
    bottom--;
    
    // bottom -> top
    for (let i = bottom - 1; i > top - 1; i--) {
      res.push(matrix[i][left]);
    }
    left++;
  }

  return res;
};

const matrix = [
  [1,2,3,4],
  [5,6,7,8],
  [9,10,11,12]
];
console.log('spiralOrder: ', spiralOrder(matrix));

File: 54. Spiral Matrix/spiralOrder.py
from typing import List

class Solution:
  def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
    left = 0
    right = len(matrix[0])
    top = 0
    bottom = len(matrix)
    res = []
    
    while left < right and top < bottom:
      for i in range(left, right):
        res.append(matrix[top][i])
      top += 1
      
      for i in range(top, bottom):
        res.append(matrix[i][right - 1])
      right -= 1
      
      if not (left < right and top < bottom):
        break
      
      for i in range(right - 1, left - 1, -1):
        res.append(matrix[bottom - 1][i])
      bottom -= 1
      
      for i in range(bottom - 1, top - 1, -1):
        res.append(matrix[i][left])
      left += 1
      
    return res
    

matrix = [
  [1, 2, 3, 4],
  [5, 6, 7, 8],
  [9, 10, 11, 12]
]
solution = Solution()
print('spiralOrder: ', solution.spiralOrder(matrix))
